subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/DependencyInsightReportTask.java

```diff
@@ -393,63 +474,112 @@ private AttributeContainer concat(AttributeContainer configAttributes, Attribute
 
-        private void writeMissingAttributes(StyledTextOutput out, AttributeContainer requested, int maxAttributeLen, Sets.SetView<Attribute<?>> missing) {
-            if (missing.size() != requested.keySet().size()) {
-                out.println();
+        private void writeAttributeTable(
+            StyledTextOutput out, AttributeContainer attributes, AttributeContainer requested, boolean selected
+        ) {
+            // Bucket attributes into three groups:
+            // 1. Attributes that are only in the variant
+            // 2. Attributes that are both in the variant and requested by the configuration
+            // 3. Attributes that are only in the requested configuration
+            List<Attribute<?>> providedAttributes = new ArrayList<>();
+            Map<Attribute<?>, AttributeMatchDetails> bothAttributes = new LinkedHashMap<>();
+            List<Attribute<?>> requestedAttributes = new ArrayList<>();
+            for (Attribute<?> attribute : attributes.keySet()) {
+                AttributeMatchDetails details = match(attribute, attributes.getAttribute(attribute), requested);
+                if (details.matchType != MatchType.NOT_REQUESTED) {
+                    bothAttributes.put(attribute, details);
+                } else {
+                    providedAttributes.add(attribute);
+                }
             }
-            out.withStyle(Description).text("      Requested attributes not found in the selected variant:");
-            out.println();
-            for (Attribute<?> missingAttribute : missing) {
-                out.withStyle(Description).text("         " + StringUtils.rightPad(missingAttribute.getName(), maxAttributeLen) + " = " + requested.getAttribute(missingAttribute));
-                out.println();
+            for (Attribute<?> attribute : requested.keySet()) {
+                // If it's not in the matches, it's only in the requested attributes
+                if (bothAttributes.values().stream().map(d -> d.requested).noneMatch(Predicate.isEqual(attribute))) {
+                    requestedAttributes.add(attribute);
+                }
             }
-        }
 
-        private void writeFoundAttributes(StyledTextOutput out, AttributeContainer attributes, AttributeContainer requested, int maxAttributeLen, Set<Attribute<?>> matchedAttributes) {
-            for (Attribute<?> attribute : attributes.keySet()) {
-                Object actualValue = attributes.getAttribute(attribute);
-                AttributeMatchDetails match = match(attribute, actualValue, requested);
-                out.withStyle(Description).text("      " + StringUtils.rightPad(attribute.getName(), maxAttributeLen) + " = " + actualValue);
-                Attribute<?> requestedAttribute = match.requested;
-                if (requestedAttribute != null) {
-                    matchedAttributes.add(requestedAttribute);
+            List<String> header = new ArrayList<>(ImmutableList.of("Name", "Provided", "Requested"));
```

Comment:
I think `Arrays.asList()` should be preferred throughout this method (and in general, as the pre-Java 11 construct a mutable list prepopulated some known values).  I thought it was weird here to use an immutable list during construction and then add to it in the next block.

Reply:
I'm not sure what you mean here. I don't think `Arrays.asList` is inherently any different from `ImmutableList.of` for initializing the lists original content. To be honest, this was probably a mistake to build it like this at all -- I think I should rewrite it to just use `ImmutableList.Builder`, which will be clearer and more efficient.